lib/value/src/kind/crud/remove.rs

```diff
@@ -0,0 +1,922 @@
+                                &mut at_path_kind
+                            }
+                            Some(child) => child,
+                        }
+                        .remove_inner(&segments[1..], compact)
+                        .compact(array, index as usize, compact)
+                    } else {
+                        // guaranteed to not delete anything
+                        CompactOptions::Never
+                    }
+                }
+                OwnedSegment::Coalesce(fields) => {
+                    let original = self.clone();
+
+                    #[allow(clippy::option_if_let_else)] // false positive due to lifetime issues
+                    if let Some(object) = self.as_object_mut() {
+                        let mut output = Self::never();
+
+                        let mut compact_options = None;
+
+                        for field in fields {
+                            let field_kind =
+                                original.at_path([OwnedSegment::Field(field.to_string())].as_ref());
+
+                            if field_kind.contains_any_defined() {
+                                let mut child_kind = original.clone();
+                                let mut child_segments = segments.to_vec();
+                                child_segments[0] = OwnedSegment::Field(field.clone());
+
+                                let child_compact_options = child_kind
+                                    .remove_inner(&child_segments, compact)
+                                    .compact(object, field.to_string(), compact);
+
+                                compact_options = Some(compact_options.map_or(
+                                    child_compact_options,
+                                    |compact_options: CompactOptions| {
+                                        compact_options.or(child_compact_options)
+                                    },
+                                ));
+                                output = output.union(child_kind);
+
+                                if !field_kind.contains_undefined() {
+                                    // No other field will be visited, so return early
+                                    break;
+                                }
+                            }
+                        }
+                        *self = output;
+                        compact_options.unwrap_or(CompactOptions::Never)
+                    } else {
+                        // guaranteed to not delete anything
+                        CompactOptions::Never
+                    }
+                }
+                OwnedSegment::Invalid => CompactOptions::Never,
+            }
+        } else {
+            CompactOptions::new(self.contains_any_defined(), self.contains_undefined())
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum CompactOptions {
+    Always,
+    Maybe,
+    Never,
+}
```

Comment:
It'd be good to add some inline comments here, even though this is an internal type.